<a href="https://colab.research.google.com/github/nagae/ICL_B_2021/blob/main/Colab_%E3%83%97%E3%83%AD%E3%82%B0%E3%83%A9%E3%83%A0%E5%85%A5%E9%96%80(1).ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Colabプログラム入門(1) 変数と四則演算

Colaboratory では， **Python** という言語を使ったプログラムが実行できる． プログラムが使えるようになることで， 複雑な科学技術計算や膨大な反復処理が可能になる．このノートブックでは，変数と四則演算について練習しよう．

# 変数と四則演算
Python においては，**変数** とは，値を格納する「箱」のようなものだ(正確には，変数は「値を格納する箱の場所」を表す)．次のセルを実行すると，`a`と`b`という2つの変数が準備され，それぞれに`10`と`6`という値が格納される．

In [1]:
## 2つの変数 a, b に値を格納する
a = 10 # a ← 10 
b = 6

変数の中身を確認するには，変数を[`print`関数](https://docs.python.org/ja/3/library/functions.html#print)の引数として渡す．

In [3]:
## 変数の中身を確認する
print(a) # a の中身だけを表示
print("a=", a) # "a="という文字列のあとに a の中身を表示
print(a, b) # a と b の中身を並べて表示
print("a=", a, "b=", b) # "a="の後に a の中身, "b=" の後に b の中身を表示

10
a= 10
10 6
a= 10 b= 6


変数 `a` と `b` に対する **四則演算** は以下のように記述できる．

In [4]:
## 四則演算
print("a+b=", a+b) # 和
print("a-b=", a-b) # 差
print("a*b=", a*b) # 積
print("a/b=", a/b) # 除
## 整数の商と剰余
print("a//b=", a//b) # 商
print("a%b=", a%b) # 剰余

a+b= 16
a-b= 4
a*b= 60
a/b= 1.6666666666666667
a//b= 1
a%b= 4


## 【練習】 四則演算と商・剰余
上述の四則演算と商・剰余を，下記の組み合わせに対して実行してみよう．

| a | b | 
|---|---|
| 2 | 5 |
| -4 | 7 |
| 11 | -3|
| -8 | -5|

In [5]:
a = 2
b = 5
print("a + b = ", a + b)
print("a - b = ", a - b)
print("a * b = ", a * b)
print("a / b = ", a / b)
print("a // b = ", a // b)
print("a % b = ", a % b)

a + b =  7
a - b =  -3
a * b =  10
a / b =  0.4
a // b =  0
a % b =  2


## 【余談】 負の数に対する商・剰余
負の数に対する商・剰余は，**プログラム言語によって異なる結果を返す**ため，極力，使わない方がよい．

例えば，同じ式であっても，C言語と呼ばれるプログラム言語と Python とでは，以下のように評価結果が異なる：

| 計算式 | C言語 | Python |
|---|---|---|
|`   11 % 3` | ` 2` | ` 2`|
|`(-11) % 3` | `-2`  | ` 1` |
|`   11 % (-3)` |` 2` | `-1` |
|`(-11) % (-3)`| `-2` | `-2` |

その理由は「[だって，そういう仕様だから](https://ja.stackoverflow.com/questions/66236/python-%E3%81%A7-%E8%B2%A0%E3%81%AE%E6%95%B0%E3%82%92%E4%BD%BF%E3%81%A3%E3%81%9F%E5%89%B0%E4%BD%99%E7%AE%97%E3%81%AB%E3%81%A4%E3%81%84%E3%81%A6)」以外には無い．文句を言っても仕方ないので，極力使わないのが賢明．

以降では，実際にC言語プログラムとPythonを実行してみて，その結果の違いを確認してみよう．



### C言語の場合
下記のセルを実行することで，C言語のソースコードが書かれた `mod.c` というファイルが生成される．

In [6]:
%%file mod.c 
#include <stdio.h>
#include <stdlib.h>
int main(int argc, char **argv){
    int a, b;
    if (argc < 2){ fprintf(stderr, "Error: insufficient argument"); return(-1); }
    else{
        a = atoi(argv[1]);
        b = atoi(argv[2]);
    }
    printf("(%d)%%(%d)=%d\n", a, b, (a%b) );
    return(a%b);
}

Writing mod.c


C言語では，`mod.c` というソースファイルはそのまま実行できず，これを**コンパイル**して実行ファイルを作る必要がある．以下のセルを実行すると，`mod.c` から実行ファイル `mod` が生成される．

In [7]:
# C言語で計算する (-11)%3, 11%(-3) および (-11)%(-3)
!gcc mod.c -o mod # C言語プログラムをコンパイルして実行ファイルを作成
!chmod 755 ./mod # 作成した実行ファイルに実行権限を与える
# 以下，それぞれ実行ファイルを呼び出して計算結果を表示
!./mod  11  3
!./mod -11  3
!./mod  11 -3
!./mod -11 -3

(11)%(3)=2
(-11)%(3)=-2
(11)%(-3)=2
(-11)%(-3)=-2


### Python の場合
Python はC言語のようなコンパイルの必要がない(**インタプリタ言語**と呼ばれる)．以下のセルを実行すれば，計算結果が得られる

In [8]:
# Pythonで計算する (-11)%3, 11%(-3) および (-11)%(-3)
print( "(11) %(3) =",    11%  3  )
print( "(-11)%(3) =", (-11)%  3  )
print( "(11) %(-3)=",    11%(-3) )
print( "(-11)%(-3)=", (-11)%(-3) )

(11) %(3) = 2
(-11)%(3) = 1
(11) %(-3)= -1
(-11)%(-3)= -2


# 百五減算
和算の一種で，江戸時代の数学の本「塵劫記」に載っているらしい．0以上105未満のある不明な数 (例えば他人の年齢) $x$ について，
- $x$ を3で割った余り $a$
- $x$ を5で割った余り $b$
- $x$ を7で割った余り $c$

のみから元の数$x$を当てる計算方法．

計算式をPython風に表せば，
```python
x = (70 * a + 21 * b + 15 * c) % 105
```
となる．

もとの和算では，これを計算するのに「括弧内が 105 を超えている場合には，105を繰返し引く」という操作をしたため **百五減算** と呼ばれている．

## 【練習】
以下のセルの `x=19` の部分を色々書き換えてみて，どんな場合にも, $a, b, c$だけから正しくもとの$x$が求められるかやってみよう．





In [12]:
#### xから a, b, c を求める
x=19 # もとの数 ← これを他の数値に変えてセルを実行してみよう
a = x % 3 # 3で割った余り
b = x % 5 # 5で割った余り
c = x % 7 # 7で割った余り
print("a=", a, "b=", b, "c=", c) # それぞれの余りを表示

#### a, b, c から x を求める
x =  (70 * a + 21 * b + 15 * c) % 105 # 百五減算でもとの数
print("x=", x)

a= 1 b= 4 c= 5
x= 19


## 【余談】
今回の講義の範疇ではないが，`for` を使うと，全ての`x`について正しく計算できるかどうかが調べられる．

In [13]:
for x in range(105): # range(1,105)は0以上105未満の整数集合を作り出す
    a = x % 3
    b = x % 5
    c = x % 7
    X = (70*a + 21*b + 15*c) % 105 # 複合化された数値
    print(x, "=", X, end=', ') # 元の数値xと複合化された数値Xを並べて表示(鬱陶しいので改行しないようにしている)

0 = 0, 1 = 1, 2 = 2, 3 = 3, 4 = 4, 5 = 5, 6 = 6, 7 = 7, 8 = 8, 9 = 9, 10 = 10, 11 = 11, 12 = 12, 13 = 13, 14 = 14, 15 = 15, 16 = 16, 17 = 17, 18 = 18, 19 = 19, 20 = 20, 21 = 21, 22 = 22, 23 = 23, 24 = 24, 25 = 25, 26 = 26, 27 = 27, 28 = 28, 29 = 29, 30 = 30, 31 = 31, 32 = 32, 33 = 33, 34 = 34, 35 = 35, 36 = 36, 37 = 37, 38 = 38, 39 = 39, 40 = 40, 41 = 41, 42 = 42, 43 = 43, 44 = 44, 45 = 45, 46 = 46, 47 = 47, 48 = 48, 49 = 49, 50 = 50, 51 = 51, 52 = 52, 53 = 53, 54 = 54, 55 = 55, 56 = 56, 57 = 57, 58 = 58, 59 = 59, 60 = 60, 61 = 61, 62 = 62, 63 = 63, 64 = 64, 65 = 65, 66 = 66, 67 = 67, 68 = 68, 69 = 69, 70 = 70, 71 = 71, 72 = 72, 73 = 73, 74 = 74, 75 = 75, 76 = 76, 77 = 77, 78 = 78, 79 = 79, 80 = 80, 81 = 81, 82 = 82, 83 = 83, 84 = 84, 85 = 85, 86 = 86, 87 = 87, 88 = 88, 89 = 89, 90 = 90, 91 = 91, 92 = 92, 93 = 93, 94 = 94, 95 = 95, 96 = 96, 97 = 97, 98 = 98, 99 = 99, 100 = 100, 101 = 101, 102 = 102, 103 = 103, 104 = 104, 

## 百五減算が機能する仕組み
百五減算のポイントは，$x$の評価式
$$
x = (70a + 21b + 15c)
$$
における$a$, $b$および$c$の係数を，それぞれ，
- $a$ の係数は，5でも7でも割り切れるが，3で割ると1余る数(この場合, 70)
- $b$ の係数は，3でも7でも割り切れるが，5で割ると1余る数(この場合, 21)
- $c$ の係数は，3でも5でも割り切れるが，7で割ると1余る数(この場合, 15)

となるように与えることである．このとき，式
$x = (70a + 21b + 15c)$は，
$$\begin{align}
x &= (70a + 21b + 15c) \\
&= a + 3(23a + 7b+5c) && \text{($x$を3で割った余りは$a$)}\\
&= b + 5(14a + 4b + 3c) && \text{($x$を5で割った余りは$b$)}\\
&= c + 7(10a + 3b + 2c) && \text{($x$を7で割った余りは$c$)}
\end{align}$$
となり，ツジツマが合うわけだ．

## 百五減算と暗号
百五減算は **暗号の一種** と捉えることもできる．もとの数$x$ から $a, b, c$を求める部分が**暗号化**に相当し，与えられた$a, b, c$からもとの数$x$ を求めるのが**複合化** に相当する．例えば，戦争中に自国の作戦コード$x$を前線にいる味方に伝える場合，メッセージとして$a, b, c$だけを送ることにすれば，たとえそのメッセージが漏れても，($x$の計算方法が伝わらない限りは)もとの$x$が相手にバレることはない．

上述の暗号・復号が正しく機能するためには，
- 暗号化($x \rightarrow a, b, c$ の計算)は，容易
- 複合化($a, b, c \rightarrow x$ の計算)は，極めて困難

という非対称性が必要．

ただし，百五減算の場合には，どんなに除数を大きくしても，計算機を使えば複合化が容易にできてしまうため，実用的な暗号にはなり得ない．

**じゃあ，実用的な暗号ってどうなってるの?** 

# RSA暗号ことはじめ
1977に米国MITの Rivest, Shamir および Adleman という3人の学者が **数理暗号** という暗号化の手法(**RSA暗号**)を発表した．この手法は「**公開鍵**」と「**秘密鍵**」という2つの情報を用いることで，**暗号化**(元の数値$x$から暗号化された値$y$を求める)の方法と**複合化**(暗号化された数値$y$から元の数値$x$を求める)の方法を**非対称にした点**が画期的である．

### RSA暗号を使うための準備
RSA暗号を使うためには，暗号の**受信者**が，以下の手順に従って**公開鍵**と**秘密鍵**を準備する必要がある．

まず，2つの大きな素数$p, q$を用意する．ここでは，簡単のために小さい素数$(p,q)=(3,11)$を使おう．この2つの素数に対して，
- $N = pq$
- $M = (p-1)(q-1)$

を計算する．上述の例では，$N=33, M = 20$である．

次に，
- $M$と**互いに素**な(i.e. $M$と$r$の最大公約数が1となる)整数 $r>1$

を求める．上述の例では $M=20$なので，$r=3$ である．この$r$ を用いて，
- $rs \mod M = 1$ となる(i.e. $rs$ を$M$で割った余りが1になる)ような整数 $s$ 

を求める．上述の例では，$s=7$ である．以上の準備のもと，暗号の**受信者**は，$(N, r)$ を **公開鍵**, $s$を **秘密鍵**とする．

公開鍵$(N, r)$は，メールで送っても構わないし，ホームページで公開しても構わない．例えば，長江の公開鍵は[こんな感じ](https://github.com/nagae.keys)で公開されている．一方，秘密鍵$s$は，決して公開してはならない．

### 暗号化と複合化
暗号の**送信者** が数値$x$を暗号化して送りたい場合，公開鍵$(N, r)$を用いた以下の式で暗号化された数値を計算する．
$$
y = x^{r} \mod N
$$
こうして暗号化された数値$y$は，$(N, r)$の情報を用いるだけでは，どうやっても解読できない(と信じられている)ため，メールとかに貼り付けるなどして受信者に送ればいい．

**受信者**は，受け取った暗号$y$に対して，秘密鍵$s$(と$N$を用いた)以下の計算を行うことで，元の数値$x$を容易に求められる．
$$
x = y^{s} \mod N
$$

Python では「べき乗」と「剰余」を求める演算子として，それぞれ，`**`および`%`が用意されているから，これを使って
```python
y = x ** r % N # もとの数→暗号コード
x = y ** s % N # 暗号コード→もとの数
```
とすることもできるが，[`pow`関数](https://docs.python.org/ja/3/library/functions.html#pow)を用いて，
```python
y = pow( x, r, N ) # もとの数→暗号コード
x = pow( y, s, N ) # 暗号コード→もとの数
```
とする方がエレガントである．

以下のセルの`x=19`の部分を書き換えて，実際に暗号化と複合化を行なってみよう．

In [14]:
(p, q) = (3, 11)
(r, s) = (3, 7)
N = p * q
print("N=", N)

### 暗号化(xから暗号化された数値yを計算)
x = 19 # もとの数値
y = pow( x, r, N ) # (x ** r) % N を求める
### 複合化(yから元の数値xを求める)
X = pow( y, s, N ) # (y ** s) % N を求める

print(x, "→", y, "→", X) # もとの値，暗号化された値，複合化された値を表示

N= 33
19 → 28 → 19


# 文字の暗号化
## Unicode による文字の数値化
RSA暗号は整数から整数への変換なので，文字を数値に直せば，やはり暗号化できる．

ここでは，日本語を含む汎用的な文字コードである [Unicode](https://ja.wikipedia.org/wiki/Unicode)の符号化方式として一般的に使われている [UTF-8](https://ja.wikipedia.org/wiki/UTF-8) を用いる方法を示そう． 

UTF-8方式では，1つ1つの文字に対して，以下のように個別の数値(コードポイント)が割り当てられており，この数値をRSA方式で暗号化することで，秘密のメッセージを送ることができる．

|文字| 数値 | 
|---|----|
|東|26481|
|北|21271|
|大|22823|
|学|23398|
|👍|128077|

ある文字に対応する数値(コードポイント)を求めるには[`ord`関数](https://docs.python.org/ja/3/library/functions.html#ord)を，数値に対応する文字を求めるには[`chr`関数](https://docs.python.org/ja/3/library/functions.html#chr)を使う．
例えば， 「東」に対応する数値を変数`x`， `128077`に対応する文字を変数 `c`に格納し，それらを表示させるには，以下のようなコードを使う．

In [18]:
x = ord("東") # 「東」に対応する数値を x に格納
c = chr(128077) # 128077 (thumbs up) に対応する文字を c に格納
print(x, c) # 数値x と文字 c を表示

26481 👍


## UTF-8 コードポイントをRSA暗号化するためには
UTFのコードポイントは21ビットのため，その値は 0 から 2097151(=$2^{21}-1$)の範囲を取り得る． RSA暗号の性質から，全ての文字を暗号化するには$N=pq$ が 2097151より大きくなるような $p$ と$q$ を選ぶ必要がある(ただし，現時点では文字が割り当てられていない数値も多いため， $(20000)_{16}=131072$ ぐらい扱えれば十分なようだ[[参考資料]](http://www.asahi-net.or.jp/~ax2s-kmtn/ref/unicode/index_u.html))．

例えば，[Prime number list](https://www.mathsisfun.com/numbers/prime-numbers-to-10k.html)から適当な素数 $(p, q) = (1051, 1999)$ を選んだとしよう．このとき，

In [19]:
(p, q) = (1051, 1999)
N = p * q
M = (p-1) * (q-1)
print("N=", N)
print("M=", M)

N= 2100949
M= 2097900


であるから，全ての文字(に対応するコードポイント)をRSA暗号化できる．

次に， 公開鍵と秘密鍵になる $r$ と $s$ を求めよう．まず，$r$は，$M = 2097900$と**互いに素となる**($M$との最大公約数が1となる)ように選べばよい． 2つの整数の最大公約数は[`math`モジュール](https://docs.python.org/ja/3/library/math.html)内の[`math.gcd`関数](https://docs.python.org/ja/3/library/math.html#math.gcd)を使えば簡単に求められるから， 以下のセルの `r=3` の部分を色々書き換えて実行すれば， `M`との最大公約数が1となるような `r` が見付けられる．


In [20]:
import math # math モジュールをインポートすることで， gcd関数が使えるようになる
r = 3
math.gcd(M, r)

3

ここでは，そのような$r$として`r=11` を採用しよう．

In [21]:
r = 11
math.gcd(M, r)

1

次に，秘密鍵$s$は，$rs$を$M$で割った余りが1になるような値である． これを手作業で求めるのはちょっと大変である．そこで，以下のチートコードを使ってしまおう：

In [22]:
# (for を使った繰返しは次回以降解説します)
# for 文を使った繰返しによって， (r*s) を M で割った余りが 1 となる s を探す
for s in range(2,M): # 2以上M未満の候補を次々と s に代入して繰返す
    if (r*s) % M ==1: # (r*s) % M が 1 となるなら, s を表示して繰返しを終了する
        print("s=", s)
        break

s= 953591


以上の手続きにより，

- **公開鍵** `(N, r) = (2100949, 11)` 
- **秘密鍵** `s = 953591`

が求められた．


この公開鍵`(N, r)`を使えば，「東北大学👍」という文章は，以下のように暗号化できる：

|文字| 数値 |  暗号コード |
|---|----| ---- |
|東|26481|1203178|
|北|21271|1163952|
|大|22823|2059934|
|学|23398|355494|
|👍|128077|1350391|

以下のようなコードにすれば，暗号化の手続きが(多少)スッキリ書けるだろう．


In [23]:
# pow と ord を組み合わせると暗号化の手続きがスッキリ書ける．
# ここでは，ord(文字)で求められた数値を pow関数の第1引数として渡している．
print( pow( ord("東"), r, N ) )
print( pow( ord("北"), r, N ) )
print( pow( ord("大"), r, N ) )
print( pow( ord("学"), r, N ) )
print( pow( ord("👍"), r, N ) )

1203178
1163952
2059934
355494
1350391


In [24]:
pow( 1203178, s, N )

26481

In [25]:
chr( 26481 )

'東'

同様にすれば，複合化の手続きも簡単に書ける．

In [26]:
# chr と pow を組み合わせるた複合化の手続き． chr と pow の順番に注意．
# ここでは，pow( 暗号化された数値, s, N ) で求めた「複合化された数値」を chr関数に渡している．
print( chr( pow( 1203178, s, N ) ) )
print( chr( pow( 1163952, s, N ) ) )
print( chr( pow( 2059934, s, N ) ) )
print( chr( pow( 355494, s, N ) ) )
print( chr( pow( 1350391, s, N ) ) )

東
北
大
学
👍


次回以降で解説する `for` 文を使えば， もっと簡単に暗号化・符号化できる.


In [29]:
# for 文を使った暗号化
text = "情報基礎B😩" # 暗号化したい文章
for c in text: # text 内の文字を順に変数 c に格納し, cのコードポイントを暗号コードに変換して表示するブロックを実行
    print( pow( ord(c), r, N ) )

# for 文を使った複合化
codes = [1997045, 1735770, 21926, 1254593, 1565615, 293423] # 暗号化されたコード(カンマで区切る)
for y in codes: # codes 内の暗号コードを順に変数 y に格納し，yを複合化したコードポイントに対応する文字を表示するブロックを実行
    print( chr( pow(y, s, N) ) )

1997045
1735770
21926
1254593
1565615
293423
情
報
基
礎
B
😩


## 【余談】 RSA暗号の鍵長
RSA暗号の$N$の鍵長としては2048ビット程度が望ましいとされている．今回の例で使った $N$ は高々20ビット程度なので，簡単に因数分解できてしまう．ちなみに，次のセルを実行すれば$2^{2048}$ がどれくらい大きいかが判るだろう．

In [30]:
print("鍵の長さ:", 2**2048)
print("「鍵の長さ」の桁数:", math.ceil(math.log10(2**2048)))

鍵の長さ: 32317006071311007300714876688669951960444102669715484032130345427524655138867890893197201411522913463688717960921898019494119559150490921095088152386448283120630877367300996091750197750389652106796057638384067568276792218642619756161838094338476170470581645852036305042887575891541065808607552399123930385521914333389668342420684974786564569494856176035326322058077805659331026192708460314150258592864177116725943603718461857357598351152301645904403697613233287231227125684710820209725157101726931323469678542580656697935045997268352998638215525166389437335543602135433229604645318478604952148193555853611059596230656
「鍵の長さ」の桁数: 617


# 課題5 三百八十五減算とRSA暗号

## 【課題5-1】三百八十五減算
百五減算より大きな数を扱うために，0以上385未満の任意の数$x$に対して，
- $x$を5で割った余り$a$
- $x$を7で割った余り$b$
- $x$を11で割った余り$c$

から$x$を求める式を求めよ．求めた式を用いて，以下のそれぞれのケースに対して，正しくもとの数$x$が復号できるか確認せよ．

|a | b | c | x |
|---|----|---|---|
|2 | 6 | 5 | 27 |
|3 | 2 | 10 | 373 |
|4 | 2 | 1 | 254 |

## 【課題5-2】 RSA暗号
$N$が少なくとも$132000$程度となるように，上述の例とは異なる素数 $p, q$を用いて，自分で公開鍵$(N, r)$と秘密鍵$s$を作成せよ． 
素数については，以下のサイトに掲載されているものを使ってよい：
https://www.mathsisfun.com/numbers/prime-numbers-to-10k.html

求めた公開鍵と秘密鍵を用いて，以下の漢詩の暗号化(漢字→暗号コード)と複合化(暗号コード→漢字)を行え：

> 国破山河在  
> 城春草木深  
> 感時花濺涙  
> 恨別鳥驚心

上述の例(`(N, r) = (2100949, 11), s = 953591`)の場合， 暗号コードは以下の通り：
```text
1649451, 1712564, 520074, 374329, 1033149
1978913, 790664, 1401088, 1127747, 631168
140273, 1827109, 814736, 266160, 1996365
1143597, 1163429, 902879, 1062966, 1560799
```

## 【おまけ】 暗号通信してみよう
Microsoft Teams の[「暗号通信」チャンネル](https://teams.microsoft.com/l/channel/19%3ad47d166eba2f4af68cd824e10920c896%40thread.tacv2/%25E6%259A%2597%25E5%258F%25B7%25E9%2580%259A%25E4%25BF%25A1?groupId=59469621-b62a-403f-8e6f-f41273ea6e2b&tenantId=f848e92a-65fa-4729-b2e6-eac667dd1bc3)を使って暗号通信してみよう．

まず，暗号を受信したい人は，チャットルームで以下のように発言する(暗号鍵は適宜変えて下さい)：

> 私の公開鍵は (2100949, 11) です． 誰か暗号を送って下さい．

メッセージを送信したい人は，その公開鍵を元に，送りたい文章の暗号コードを，以下のようにして送る：

 > ◯◯さん，暗号を送ります: 1576653, 174297, 538933, 
715971, 1655843, 1509974, 1689836, 524070

暗号コードをもらったら，秘密鍵を使って複合化し，**リアクション**しよう．

> □□さん，ありがとう．頑張ろうね!

もちろん暗号化されているからといって，公序良俗に反するメッセージやクレジット・カード番号などの金融情報は送ってはいけない．また，[【余談】](#scrollTo=GgFZYwryFDjs&line=1&uniqifier=1)にも書いたように，そもそも7〜8桁程度の鍵長では，大して苦もなく因数分解できてしまうことにも注意されたい．